java-frontend/src/main/java/org/sonar/java/se/ProgramState.java

```diff
@@ -179,12 +180,22 @@ public String toString() {
 
   public ProgramState addConstraint(SymbolicValue symbolicValue, Constraint constraint) {
     PMap<SymbolicValue, Constraint> newConstraints = constraints.put(symbolicValue, constraint);
-    if (newConstraints != constraints) {
+    if (newConstraints != constraints && !isSymbolicValueOfVolatileField(symbolicValue)) {
       return new ProgramState(this, newConstraints);
     }
     return this;
   }
 
+  private boolean isSymbolicValueOfVolatileField(SymbolicValue symbolicValue) {
```

Comment:
What is tested here is that there are no volatile symbol that can reach this SV whereas this is not exactly what the ticket is supposed to solve : the ticket is about a specific symbol that is volatile and thus its SV may have changed (but not the constraints on the currently assigned SV).